Client/src/hooks/dataRestriction.ts

```diff
@@ -16,14 +19,148 @@ import {
+  label
+} from 'ebrc-client/App/DataRestriction/DataRestrictionActionCreators';
 
 const STUDY_ACTION_CLASS_NAME = 'study-action';
 
 const STUDY_ID_DATA_ATTRIBUTE = 'data-study-id';
 const ARGS_DATA_ATTRIBUTE = 'data-args';
 
+export type CompleteApprovalStatus = 'approved' | 'not-approved' | 'study-not-found';
+
+export type ApprovalStatus = CompleteApprovalStatus | 'loading';
+
+/**
+ * @param studyId
+ * @param action
+ * @returns An ApprovalStatus signifying if the user is permitted to
+ * perform the specified "action" for the study with id "studyId"
+ */
+export function useApprovalStatus(studyId: string, action: string) {
+  const [approvalStatus, setApprovalStatus] = useState<ApprovalStatus>(
+    'loading'
+  );
+
+  const attemptActionTask = useAttemptActionTask(studyId, action);
+
+  useEffect(() => {
+    setApprovalStatus('loading');
+
+    return attemptActionTask.run(setApprovalStatus);
+  }, [attemptActionTask]);
+
+  return approvalStatus;
+}
+
+/**
+ * @param studyId
+ * @param action
+ * @param onApprovalStatusCheckComplete
+ * @returns A callback, which, when executed:
+ *
+ * (1) attempts to perform "action" for the study with "studyId"
+ *
+ * (2) executes "onApprovalStatusCheckComplete" once the user's permission to
+ * perform the action has been checked
+ *
+ * NOTE: whenever the passed "studyId", "action", or "onApprovalStatusCheckComplete"
+ * arguments have changed, or the component which invoked this hook has been unmounted,
+ * all pending executions of "onApprovalStatusCheckComplete" will be canceled
+ */
+export function useScopedAttemptActionCallback(
+  studyId: string,
+  action: string,
+  onApprovalStatusCheckComplete: (
+    approvalStatus: CompleteApprovalStatus
+  ) => void
+) {
+  const attemptActionTask = useAttemptActionTask(studyId, action);
+
+  const cancelCbs = useRef<Set<() => void>>(new Set());
+
+  const attemptActionCallback = useCallback(() => {
+    const cancelCb = attemptActionTask.run(onApprovalStatusCheckComplete);
+    cancelCbs.current.add(cancelCb);
+  }, [attemptActionTask, onApprovalStatusCheckComplete]);
+
+  useEffect(
+    () => () => {
+      for (const cancelCb of cancelCbs.current) {
+        cancelCb();
+      }
+
+      cancelCbs.current.clear();
+    },
+    [attemptActionCallback]
+  );
+
+  return attemptActionCallback;
+}
+
+/**
+ * @param studyId
+ * @param action
+ * @returns A Task which, when run, will:
+ * (1) check if the user is permitted to perform "action" for the study with id "studyId"
+ * (2) if appropriate, open a Data Restriction modal
+ * (3) resolve to the appropriate ApprovalStatus
+ */
+export function useAttemptActionTask<E>(
+  studyId: string,
+  action: string
+): Task<CompleteApprovalStatus, E> {
+  const dispatch = useDispatch();
+  const wdkDependencies = useContext(WdkDepdendenciesContext);
+
+  if (wdkDependencies == null) {
+    throw new Error(
+      'In order to "useAttemptActionTask", WdkDependenciesContext must be provided.'
+    );
+  }
```

Comment:
I think it might be time to abstract this. We have a similar, generalized pattern [here](https://github.com/VEuPathDB/web-eda/blob/main/src/lib/core/hooks/nonNullableContext.ts).

Reply:
Yeah, this abstracting is overdue. Guarding against nil `WdkDependenciesContext` is a common source of duplicate code.

What do you think about moving `useNonNullableContext<T>` over to `WDKClient`? Doing so might also be a good opportunity for me to [finally] fix the `WdkDepdendenciesContext` typo.

Reply:
I think moving this to `WdkClient` is a good idea. Feel free to do this, or I can in a little bit.

Reply:
I'll be offline for a few hours, so if you have the bandwidth now, feel free to go for it! If not, I can take care of it later tonight. Let me know either way.

Reply:
https://github.com/VEuPathDB/WDKClient/pull/139